Add GH Gauge base class and Harmonic derived class

nilsdeppe opened the pull request:
## Proposed changes

This will allow us to choose a variety of gauge conditions more easily from the input file and will make GH+GRMHD FD solvers easier to write.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add GH Gauge base class and Harmonic derived class

knelli2 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Harmonic.hpp:
```diff
@@ -0,0 +1,59 @@
+class Harmonic final : public GaugeCondition {
+ public:
+  using options = tmpl::list<>;
+
+  static constexpr Options::String help{"Apply the Harmonic gauge condition."};
```
Add the `H_a = 0` part to the help string as well.

knelli2 reviewed (changes requested):
Squash immediately.

Will a future PR be changing damped harmonic into this base class format as well?

nilsdeppe commented:
Re DH: Yes, I have an analytic and DH class coded up. My plan is:
- analytic PR
- DH PR
- tags PR
- time derivative and compute tag PR
Last 2 I still need to finish

Rebased and squashed. Thanks for the review!

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/RegisterDerived.cpp:
```diff
@@ -0,0 +1,21 @@
+void register_derived_with_charm() {
+  impl<1>();
+  impl<2>();
+  impl<3>();
```
Not sure I understand the purpose of these...are these just creating "gauges" that are empty?

nilsdeppe replied:
Plan is to have dim-dependent gauges (e.g. have Analytic also register the analytic conditions that can be imposed)

Edit: that will be the next PR

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Gauges.cpp:
```diff
@@ -0,0 +1,8 @@
+
+namespace GeneralizedHarmonic::gauges {
+GaugeCondition::GaugeCondition(CkMigrateMessage* msg) : PUP::able(msg) {}
+}  // namespace GeneralizedHarmonic::gauges
```
Just so I understand, I wonder why you need this file at all?

nilsdeppe replied:
I want to define the constructor in a `cpp` file instead of an `hpp`

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Gauges.hpp:
```diff
@@ -3,7 +3,32 @@
+namespace gauges {
+/// \brief Base class for GH gauge conditions.
+///
+/// Derived class must have:
```
Incomplete documentation?

nilsdeppe replied:
I've rebased and squashed, updated the docs. However, the docs is written for the final state on my branch, so there are some functions and type aliases discussed that won't yet be merged. This just makes managing the PRs and commits easier on my end, and nobody can use this until all features are in anyway :)

geoffrey4444 commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Test_Harmonic.cpp:
```diff
@@ -0,0 +1,61 @@
+  tnsr::ab<DataVector, Dim, Frame::Inertial> d4_gauge_h(num_points);
+
+  const double time = std::numeric_limits<double>::signaling_NaN();
+  const tnsr::I<DataVector, Dim, Frame::Inertial> inertial_coords(num_points);
```
Is this automatically initialized to signaling nans?

nilsdeppe replied:
Yes

geoffrey4444 reviewed (dismissed):
Please consider my comments before merging, but they are optional

nilsdeppe commented:
Rebased to force GitHub to run actions. GitHub was having issues with Actions yesterday, and sometimes they wouldn't start. No code changes since last review.

kidder commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Gauges.hpp:
```diff
@@ -3,7 +3,37 @@
+/// condition. The `GeneralizedHarmonic::gauges::dispatch()` function must also
+/// be updated to correctly detect and forward to the gauge condition. The
+/// header file must also be included in `Factory.hpp` and the gauge condition
+/// added to the `all_gauges` type alias in `Factory.hpp`.
```
I don't see this type alias anywhere

kidder commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Harmonic.cpp:
```diff
@@ -0,0 +1,58 @@
+
+void Harmonic::pup(PUP::er& p) { GaugeCondition::pup(p); }
+
+std::unique_ptr<GaugeCondition> Harmonic::get_clone() const {
```
`#include <memory>`

kidder reviewed (changes requested):
you can squash

nilsdeppe commented:
Rebased and squashed. Thanks for the reviews!

kidder commented:
ignored test timeouts
